[REVIEW] ARIMA: follow cuML array conventions and remove redundancy

Nyrio opened the pull request:
Follow-up on a discussion in #2928

I simply changed the `get_params` and `set_params` to a property `params` to avoid overriding functions from the base class with a different meaning (i.e fit parameters instead of init parameters).

@JohnZed What are your thoughts about also having `get_params` and `set_params` in ARIMA (with the sklearn meaning)? There is the inconvenience that for time series models, the dataset is an init param, and it would also require moving some code from `__init__` to `fit`, but I don't have a strong opinion on whether we should have these functions. (also, it will likely require more than just overriding `get_param_names` because the order is stored in a C structure instead of separate class attributes)

GPUtester commented:
Can one of the admins verify this patch?

GPUtester commented:
Can one of the admins verify this patch?

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

mdemoret-nv reviewed (changes requested):
I have a few comments:

1. Personally, I would prefer to use a different name than `params`. I think it's too close to `get_params` and its not clear how `get_params` differs from the property `params`. Something like `get_fit_params` or `solver_params` where we distinguish what these parameters belong to would be better. Additionally, the difference should be explained in the function docstring.
2. I think converting `get_params` to a `@property` could lead to bad behavior by the user and unexplained behavior. For example, the user could do `arima.params["mu"] = 5.0` which would work perfectly but would not have the desired effect. I think its much clearer to the user that this will not have the desired effect `arima.params()["mu"] = 5.0`. Keeping this as a `set/get` functions aligns more with its behavior and would prevent weird scenarios like: 
```python
arima = ARIMA()
if "sigma2" not in arima.params:
    arima.params = {"sigma2": 10}
print(arima.params["sigma2"]) # Would raise an exception
```
3. By changing the name of `get_params`/`set_params` this exposes the base functionality which will return a different dictionary. Either we need to support the `Base.get_params()` functionality, or warn/prevent the user from using it.

mdemoret-nv commented:
> @JohnZed What are your thoughts about also having `get_params` and `set_params` in ARIMA (with the sklearn meaning)? There is the inconvenience that for time series models, the dataset is an init param, and it would also require moving some code from `__init__` to `fit`, but I don't have a strong opinion on whether we should have these functions. (also, it will likely require more than just overriding `get_param_names` because the order is stored in a C structure instead of separate class attributes)

Personally, I would prefer to have `get_params` and `set_params` in ARIMA follow the sklearn meaning. Otherwise it will be the only estimator which does not. I'm not entirely sure how difficult it would be to support it, but there are a few tricks to make the base functionality work with `get_param_names` by adding properties that marshal the data to the right location. For example, I did something very similar with parameters in a C structure for `RandomProjection` here: https://github.com/rapidsai/cuml/pull/2928/files#diff-77b0e91caa369a50e8c3aaf207c1f239R179-R225

JohnZed reviewed (dismissed):
Looks good!

JohnZed commented:
> I have a few comments:
> 
> 1. Personally, I would prefer to use a different name than `params`. I think it's too close to `get_params` and its not clear how `get_params` differs from the property `params`. Something like `get_fit_params` or `solver_params` where we distinguish what these parameters belong to would be better. Additionally, the difference should be explained in the function docstring.
> 2. I think converting `get_params` to a `@property` could lead to bad behavior by the user and unexplained behavior. For example, the user could do `arima.params["mu"] = 5.0` which would work perfectly but would not have the desired effect. I think its much clearer to the user that this will not have the desired effect `arima.params()["mu"] = 5.0`. Keeping this as a `set/get` functions aligns more with its behavior and would prevent weird scenarios like:
> 
> ```python
> arima = ARIMA()
> if "sigma2" not in arima.params:
>     arima.params = {"sigma2": 10}
> print(arima.params["sigma2"]) # Would raise an exception
> ```
> 
> 1. By changing the name of `get_params`/`set_params` this exposes the base functionality which will return a different dictionary. Either we need to support the `Base.get_params()` functionality, or warn/prevent the user from using it.

Is `.params` used here the same way as in statsmodels?

JohnZed commented:
> 2. I think converting `get_params` to a `@property` could lead to bad behavior by the user and unexplained behavior. For example, the user could do `arima.params["mu"] = 5.0` which would work perfectly but would not have the desired effect. I think its much clearer to the user that this will not have the desired effect `arima.params()["mu"] = 5.0`. Keeping this as a `set/get` functions aligns more with its behavior and would prevent weird scenarios like:


Ah, the assignment issue raised here is a meaninful one we should address... would the statsmodels API have the same risk?

Nyrio commented:
> Personally, I would prefer to have `get_params` and `set_params` in ARIMA follow the sklearn meaning. Otherwise it will be the only estimator which does not. I'm not entirely sure how difficult it would be to support it, but there are a few tricks to make the base functionality work with `get_param_names` by adding properties that marshal the data to the right location. For example, I did something very similar with parameters in a C structure for `RandomProjection` here: https://github.com/rapidsai/cuml/pull/2928/files#diff-77b0e91caa369a50e8c3aaf207c1f239R179-R225

Time series estimators are also the only ones -I believe- to follow an API that is not part of sklearn. And the only ones where the model object contains the dataset, which is necessary because in time series forecasting there is no distinction between the training set and the inference set, the only dataset is the past, and what we infer is the future.

If I understand correctly, these functions in sklearn are meant to facilitate hyperparameter search. With `ARIMA`, assuming that someone is making their own hyperparameter search instead of using `AutoARIMA`, they could benefit from a mechanism to change the parameters without re-creating the cuML array every time (expensive if they pass a host data structure). It's not hard to provide this functionality, mostly refactoring.

Tl; dr: I agree that `get_params` and `set_params` could be useful with appropriate changes in the model.

> Personally, I would prefer to use a different name than `params`. I think it's too close to `get_params` and its not clear how `get_params` differs from the property `params`. Something like `get_fit_params` or `solver_params` where we distinguish what these parameters belong to would be better. Additionally, the difference should be explained in the function docstring.

Regarding the name, statsmodels' `SARIMAXResults` has a `params` property that contains the fit parameters, and in the ARIMA literature parameters unambiguously refer to the AR, MA, etc params. What we call parameters in the sklearn mindset are called the ARIMA order. I understand the need for consistency, but in this case we have conflicting references, so we have to choose arbitrarily to differ from one or the other.

> Is `.params` used here the same way as in statsmodels?

Yes and no. It contains the same data but not in the same format. Statsmodels returns a packed vector of the concatenated parameters, whereas we return a dictionary of vectors. We have a way to get the dense vector of parameters, with the method `pack`. And statsmodels has a way to get individual parameters vectors with separate properties `arparams`, `seasonalarparams` etc.

The proper way to save model parameters in our `ARIMA` model is to use `pack` and `unpack`, but getting the parameters as a dictionary of arrays can be useful for users, which was my original idea when providing these functions. I am open to ideas regarding this interface and how we could improve it.

At the moment the parameters can also be accessed with their individual attributes (e.g `self.ar`), but we don't advertise that because the plan is to eventually move the optimizer to C++ so these attributes shouldn't exist anymore (they will likely be replaced by either a packed device array, or separate device arrays, whichever requires less packing/unpacking for the optimizer) unless we decide to interface them in Python (property or automatic `self._smth` conversion).

> I think converting `get_params` to a `@property` could lead to bad behavior by the user and unexplained behavior. For example, the user could do `arima.params["mu"] = 5.0` which would work perfectly but would not have the desired effect. I think its much clearer to the user that this will not have the desired effect `arima.params()["mu"] = 5.0`. Keeping this as a `set/get` functions aligns more with its behavior and would prevent weird scenarios like:
> 
> 
> ```python
> arima = ARIMA()
> if "sigma2" not in arima.params:
>     arima.params = {"sigma2": 10}
> print(arima.params["sigma2"]) # Would raise an exception
> ```

This is a fair point.

> By changing the name of `get_params`/`set_params` this exposes the base functionality which will return a different dictionary. Either we need to support the `Base.get_params()` functionality, or warn/prevent the user from using it.

If I understand correctly, you are of the opinion that we should have this functionality. Otherwise we have the option of overriding and raising an exception.

@mdemoret-nv @JohnZed It seems that we could benefit from a discussion about the `ARIMA` interface and in particular how to keep it consistent with both statsmodels and other cuML models, and as easy to use as possible. If you wish to schedule an offline discussion to look at those APIs and decide which way to go, please let me know.

mdemoret-nv commented:
@Nyrio Brainstorming some ideas here. Since the goal is to match the functionality of both statsmodels and Sklearn, while preventing the user from using the properties incorrectly, what do you think of this plan?

1. Convert the `params` getter into two separate read-only properties `arparams`, and `seasonalparams`
   1. (I only briefly looked over the statsmodels code, so these may not be the correct property names)
   1. The goal here is interoperability with statsmodels. You mentioned the data was the same but the format was different. The goal would be to match the format but make them read only
2. Change the `params` setter into a separate method called `set_order()`
   1. This is based off your comment that these parameters could be called the "ARIMA order"
   1. I agree with you that the functionality you have in the `params` setter right now could be useful to users. If we change it to a method, this eliminates the weird use case of having a property that returns a dictionary and makes the functionality more explicit.
3. Raise exceptions in `get_params` and `set_params`
   1. Given the short time frame before 0.16, this would just prevent the base implementation since it wont work. In the future, we can still refactor to add this functionality. This will give a more meaningful error message

Let me know your thoughts.

Nyrio commented:
@mdemoret-nv Sorry for the late answer, I have been delayed by a bug. I have updated the PR, I will explain some details below.

> Convert the `params` getter into two separate read-only properties `arparams`, and `seasonalparams`  
> i. (I only briefly looked over the statsmodels code, so these may not be the correct property names)
> ii. The goal here is interoperability with statsmodels. You mentioned the data was the same but the format was different. The goal would be to match the format but make them read-only

I agree that we should have read-only properties for these.
I would like to keep the current names, though, even if it means differing a bit from statsmodels. The reasons are:
- Consistency (these names are all over the codebase)
- Simplicity and readability: `sar` vs `seasonalarparams` (verbose name without snake_case...)

I have checked the conventions that have been put in place after this model was written and iiuc the fit parameters should be cuML arrays with a leading and trailing underscore, and the properties will be implicitely defined.

The current implementation is using numpy arrays because of the scipy-based optimizer. It is ok to use cuML arrays because most functions copy the parameters to the device anyway. With this change, `fit` will still work with numpy arrays, but what is stored in the model will be cuML arrays.

> Change the `params` setter into a separate method called `set_order()`
> i. This is based off your comment that these parameters could be called the "ARIMA order"
> ii. I agree with you that the functionality you have in the `params` setter right now could be useful to users. If we change it to a method, this eliminates the weird use case of having a property that returns a dictionary and makes the functionality more explicit.

Maybe my previous message wasn't clear, but the order refers to `(p,d,q)(P,D,Q,s)`, i.e some arguments of the constructor (along with the dataset and a few others).

So a setter that provides the same functionality that we currently have would be `set_fit_params`. I have added it, with a sentence in the docstring indicating that it's different from `set_params`. I have also added `get_fit_params`, with the same functionality as before, except that it's using `output_type` instead of always returning numpy arrays.

> Raise exceptions in `get_params` and `set_params`
> Given the short time frame before 0.16, this would just prevent the base implementation since it won't work. In the future, we can still refactor to add this functionality. This will give a more meaningful error message

Ok, agreed. I have overriden the functions to raise `NotImplementedError`


By the way, it is unrelated but I encountered this issue: the assertions that enforce naming conventions don't make the difference between semi-private and private attributes:
```
AssertionError: Invalid Public Array-Like Attribute!
Public array-like attributes should always be CumlArray. Attribute: '__ar'
```
And paradoxically I don't think that there is any assertion preventing the use of a non-cuML array-like attribute if its name doesn't have a leading or trailing underscore.

Anyway, I switched to cuML arrays so this is fixed.

teju85 commented on python/cuml/tsa/arima.pyx:
```diff
@@ -406,25 +402,33 @@ class ARIMA(Base):
+                                                check_dtype=np.float64)
+                setattr(self, "_{}_".format(param_name), array)
+
+    def get_params():
+        raise NotImplementedError
+
+    def set_params():
+        raise NotImplementedError
```
It is better to instead overwrite `get_param_names()` and raise an exception there.

mdemoret-nv commented on notebooks/arima_demo.ipynb:
```diff
@@ -209,8 +209,24 @@
    "source": [
-    "param_mig = model_mig.get_params()\n",
-    "print(param_mig.keys())"
+    "print(model_mig.get_fit_params()[\"ma\"])\n",
```
Deleting the object `param_mig` causes an exception in the next cell

mdemoret-nv commented on python/cuml/tsa/arima.pyx:
```diff
@@ -388,15 +384,15 @@ class ARIMA(Base):
-        --------
-        params: Dict[str, np.ndarray]
+        -------
             A dictionary of parameter names and associated arrays
```
Missing the return object name and type.

mdemoret-nv commented on python/cuml/tsa/arima.pyx:
```diff
@@ -406,25 +402,30 @@ class ARIMA(Base):
         ----------
-        params: Mapping[str, np.ndarray]
-            A mapping (e.g dictionary) of parameter names and associated arrays
+        params:
```
Why did you remove the type in the docstring?

mdemoret-nv commented on python/cuml/tsa/arima.pyx:
```diff
@@ -880,58 +919,57 @@ class ARIMA(Base):
         if k > 0:
-            params["mu"] = np.array(x_mat[0], order='F')
+            mu_temp = np.array(x_mat[0], order='F')
+            self._mu_, *_ = input_to_cuml_array(mu_temp)
         if p > 0:
-            params["ar"] = np.array(x_mat[k:k+p], order='F')
+            ar_temp = np.array(x_mat[k:k+p], order='F')
+            self._ar_, *_ = input_to_cuml_array(ar_temp)
         if q > 0:
-            params["ma"] = np.array(x_mat[k+p:k+p+q], order='F')
+            ma_temp = np.array(x_mat[k+p:k+p+q], order='F')
+            self._ma_, *_ = input_to_cuml_array(ma_temp)
         if P > 0:
-            params["sar"] = np.array(x_mat[k+p+q:k+p+q+P], order='F')
+            sar_temp = np.array(x_mat[k+p+q:k+p+q+P], order='F')
+            self._sar_, *_ = input_to_cuml_array(sar_temp)
         if Q > 0:
-            params["sma"] = np.array(x_mat[k+p+q+P:k+p+q+P+Q], order='F')
-        params["sigma2"] = np.array(x_mat[k+p+q+P+Q], order='F')
-
-        self.set_params(params)
+            sma_temp = np.array(x_mat[k+p+q+P:k+p+q+P+Q], order='F')
+            self._sma_, *_ = input_to_cuml_array(sma_temp)
+        sigma2_temp = np.array(x_mat[k+p+q+P+Q], order='F')
+        self._sigma2_, *_ = input_to_cuml_array(sigma2_temp)
```
For reference, `input_to_cuml_array` can convert the order and dtype in a single step and `CumlArray` is capable of slicing. So you can reduce much of this code. For example, `x_mat` can be created with:
```python
x_mat = input_to_cuml_array(np.reshape(x, (N, self.batch_size)), order='F', convert_to_dtype=np.float64).array
```
Then only 1 data transfer happens to the device and all slicing happens on the GPU
Note: Adding `.array` to the end of `input_to_cuml_array` eliminates the need to do `x_mat, _, *_ = input_to_cuml_array()`. And then you can just create the individual attributes with slicing like normal:
```python
self._mu_ = x_mat[0]
self._ar_ = x_mat[k:k+p]
...
```

Nyrio replied:
Slicing doesn't create a contiguous array, it just returns the same array with different offsets and strides.
I went for another solution: interface the CUDA versions of `pack` and `unpack` (so only one host <-> device copy).

mdemoret-nv replied:
Sounds good to me.

mdemoret-nv commented on python/cuml/tsa/arima.pyx:
```diff
@@ -320,22 +322,16 @@ class ARIMA(Base):
+            d_ar_ptr = self._ar_.ptr
         if order_kf.q:
-            d_ma, *_ = input_to_cuml_array(self.ma, check_dtype=np.float64)
-            d_ma_ptr = d_ma.ptr
+            d_ma_ptr = self._ma_.ptr
         if order_kf.P:
-            d_sar, *_ = input_to_cuml_array(self.sar, check_dtype=np.float64)
-            d_sar_ptr = d_sar.ptr
+            d_sar_ptr = self._sar_.ptr
         if order_kf.Q:
-            d_sma, *_ = input_to_cuml_array(self.sma, check_dtype=np.float64)
-            d_sma_ptr = d_sma.ptr
-        d_sigma2, *_ = input_to_cuml_array(self.sigma2, check_dtype=np.float64)
-        d_sigma2_ptr = d_sigma2.ptr
+            d_sma_ptr = self._sma_.ptr
+        d_sigma2_ptr = self._sigma2_.ptr
 
         cdef ARIMAParams[double] cpp_params
         cpp_params.mu = <double*> d_mu_ptr
```
Building the `ARIMAParams` seems to happen several times (I think 4?) in this class. Might want to make an internal method that can be reused

Nyrio replied:
I don't think it can work with an internal method because `ARIMAParams` is a C++ type (as it's a template structure) so a Python function can't return it. What I went for after a bit of research is a wrapper Cython class around `ARIMAParams`. If you know a cleaner solution I'd be curious to hear about it.

mdemoret-nv replied:
An internal method can return a C++ type if its defined as a `cdef` function and only accessed from cython code. However, I think the cython wrapper works well too.

Nyrio replied:
If I remember correctly, a non-`cdef` class can't have a `cdef` method, and the class here can't be `cdef`ed because it inherits from a Python class.

mdemoret-nv commented on python/cuml/tsa/arima.pyx:
```diff
@@ -863,7 +844,65 @@ class ARIMA(Base):
+        if order_kf.k:
+            d_mu_ptr = self._mu_.ptr
+        if order_kf.p:
+            d_ar_ptr = self._ar_.ptr
+        if order_kf.q:
+            d_ma_ptr = self._ma_.ptr
+        if order_kf.P:
+            d_sar_ptr = self._sar_.ptr
+        if order_kf.Q:
+            d_sma_ptr = self._sma_.ptr
+        d_sigma2_ptr = self._sigma2_.ptr
+
+        cdef ARIMAParams[double] cpp_params
+        cpp_params.mu = <double*> d_mu_ptr
+        cpp_params.ar = <double*> d_ar_ptr
+        cpp_params.ma = <double*> d_ma_ptr
+        cpp_params.sar = <double*> d_sar_ptr
+        cpp_params.sma = <double*> d_sma_ptr
+        cpp_params.sigma2 = <double*> d_sigma2_ptr
+
+        cdef uintptr_t d_y_kf_ptr = \
+            self._d_y_diff.ptr if self.simple_differencing else self._d_y.ptr
+
+        n_obs_kf = (self.n_obs_diff if self.simple_differencing
+                    else self.n_obs)
+
+        cdef LoglikeMethod ll_method = MLE
+        diff = self.simple_differencing
+
+        d_vs = cumlArray.empty((n_obs_kf, self.batch_size), dtype=np.float64,
+                               order="F")
+        cdef uintptr_t d_vs_ptr = d_vs.ptr
+
+        batched_loglike(handle_[0], <double*> d_y_kf_ptr,
+                        <int> self.batch_size, <int> n_obs_kf, order_kf,
+                        cpp_params, <double*> vec_loglike.data(),
+                        <double*> d_vs_ptr, <bool> False, <bool> True,
+                        ll_method, <int> 0)
+
+        return np.array(vec_loglike, dtype=np.float64)
```
Unclear why `llf` changed so much here. Whats the reasoning for not using `self._loglike()` anymore?

Nyrio replied:
`_loglike()` and `_loglike_grad()` are primarily used by the scipy interpreter so they take a numpy parameter array, which is then copied to device and unpacked. `llf` is using the final parameters which are now stored as cuML arrays, so already unpacked and on device. It wouldn't make sense to copy the parameters from device to host then device again.

What I could do to remove redundancy, however, is to create a third function similar to the current `llf`, that takes the unpacked device arrays, and base both functions on this third one by moving the unpacking of `_batched_loglike` to the Python side. But this won't really be much more elegant than the current solution, and certainly not faster, so at this point I would just leave it like this, and the redundancy will disappear when this code moves to the C++ side.

mdemoret-nv replied:
Not sure i followed all of that, but I'll take your word for it.

mdemoret-nv reviewed (changes requested):
Wanted to finish this review while its fresh in my head even though it's moving to 0.17. Couple of things that can be cleaned up including the example notebook which is raising an error.

Additionally, I had a couple of notes on using `CumlArray` to simplify the code and make it more efficient.

teju85 reviewed (approved):
Changes LGTM.
